internal/match/dict.go

```diff
@@ -0,0 +1,405 @@
+	if !ok {
+		return BadWord
+	}
+	return id
+}
+
+// Words returns the current word list.
+// The list is not a copy; the caller can read but must not modify the list.
+func (d *Dict) Words() []string {
+	return d.list
+}
+
+// A Word represents a single word found in a text.
+type Word struct {
+	ID WordID
+	Lo int32 // word appears at text[Lo:Hi]
+	Hi int32
+}
+
+// InsertSplit splits text into a sequence of lowercase words,
+// inserting any new words in the dictionary.
+func (d *Dict) InsertSplit(text string) []Word {
+	return d.split(text, true)
+}
+
+// Split splits text into a sequence of lowercase words.
+// It does not add any new words to the dictionary.
+// Unrecognized words are reported as having ID = BadWord.
+func (d *Dict) Split(text string) []Word {
+	return d.split(text, false)
+}
+
+func (d *Dict) split(text string, insert bool) []Word {
+	var wbuf []byte
+	var words []Word
+	t := text
+	for t != "" {
+		var w []byte
+		var lo, hi int32
+		{
+			switch t[0] {
+			case '<':
+				if size := htmlTagSize(t); size > 0 {
+					t = t[size:]
+					continue
+				}
+			case '{':
+				if size := markdownAnchorSize(t); size > 0 {
+					t = t[size:]
+					continue
+				}
+			case '&':
+				// Assume HTML entity is punctuation.
+				if size := htmlEntitySize(t); size > 0 {
+					if t[:size] == "&copy;" {
+						lo = int32(len(text) - len(t))
+						hi = lo + int32(size)
+						w = copyright
+						t = t[size:]
+						goto Emit
+					}
+					t = t[size:]
+					continue
+				}
+			}
+			if len(t) >= 2 && t[0] == ']' && t[1] == '(' {
+				if size := markdownLinkSize(t); size > 0 {
+					t = t[size:]
+					continue
+				}
+			}
+
+			r, size := utf8.DecodeRuneInString(t)
+			if !isWordStart(r) {
+				t = t[size:]
+				continue
+			}
+			wbuf = appendFoldRune(wbuf[:0], r)
+
+			// Scan whole word.
+			lo = int32(len(text) - len(t))
+			if r != '©' {
+				for size < len(t) {
+					r, s := utf8.DecodeRuneInString(t[size:])
+					if !isWordContinue(r) {
+						break
+					}
+					size += s
+					wbuf = appendFoldRune(wbuf, r)
+				}
+			}
+			hi = lo + int32(size)
+			t = t[size:]
+
+			w = wbuf
+
+			// Special case rewrites suggested by SPDX.
+			switch {
+			case string(w) == "https":
+				// "https" -> "http".
+				w = w[:4]
+
+			case string(w) == "c" && lo > 0 && text[lo-1] == '(' && int(hi) < len(text) && text[hi] == ')':
+				w = copyright
+				lo--
+				hi++
+
+			case string(w) == "©":
+				w = copyright
+			}
+		}
+
+	Emit:
+		id, ok := d.dict[string(w)]
+		if ok {
+			if len(words) > 0 && words[len(words)-1].ID == id && string(w) == "copyright" {
+				// Treat "Copyright ©" as a single "copyright" instead of two.
+				continue
+			}
+			words = append(words, Word{id, lo, hi})
+			continue
+		}
+
+		if insert {
+			words = append(words, Word{d.Insert(string(w)), lo, hi})
+			continue
+		}
+
+		// Unknown word
+		words = append(words, Word{BadWord, lo, hi})
+	}
+
+	return words
+}
+
+var copyright = []byte("copyright")
```

Comment:
move this above the code that uses it, and comment its purpose. it's too special purpose not to be explained.